p2p/gossip/protocol.go

```diff
@@ -216,58 +252,74 @@ func (s *Neighborhood) getMorePeers(numpeers int) {
 	nds := s.ps.SelectPeers(numpeers)
 	ndsLen := len(nds)
 	if ndsLen == 0 {
-		go func() {
-			time.Sleep(1 * time.Second)
-			s.morePeersReq <- struct{}{}
-		}()
+		s.Debug("Peer sampler returned nothing.")
 		// this gets busy at start so we spare a second
-		return // we cant connect if we don't have peers
+		time.Sleep(NoResultsTimeout)
+		s.getMorePeers(numpeers)
+		return // zero samples here so no reason to proceed
 	}
 
 	// Try a connection to each peer.
 	// TODO: try splitting the load and don't connect to more than X at a time
 	for i := 0; i < ndsLen; i++ {
 		go func(nd node.Node, reportChan chan cnErr) {
-			spew.Dump(nd)
 			c, err := s.cp.GetConnection(nd.Address(), nd.PublicKey())
 			reportChan <- cnErr{nd, c, err}
 		}(nds[i], res)
 	}
 
 	i, j := 0, 0
-	for cne := range res {
-		i++ // We count i everytime to know when to close the channel
+	tm := time.NewTimer(time.Second * 20 * time.Duration(ndsLen))
```

Comment:
since you're calling GetConnection in parallel the timeout shouldn't be multiply by ndsLen